src/mots/directory.py

```diff
@@ -0,0 +1,254 @@
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+"""Directory classes for mots."""
+
+from __future__ import annotations
+from collections import defaultdict
+from dataclasses import asdict
+from dataclasses import dataclass
+from dataclasses import InitVar
+import logging
+from mots.bmo import BMOClient
+from mots.module import Module
+from mots.config import FileConfig
+from mots.utils import parse_real_name
+
+logger = logging.getLogger(__name__)
+
+WIKI_TMPL = """
+{% macro people_entry(people) %}
+{%- for p in people %}{{ p.nick }}{% if not loop.last %}, {% endif %}{% endfor %}
+{% endmacro %}
+
+{%- macro module_entry(module) -%}
+{% raw %}{{Module{% endraw %}
+|name={{ module.name }}
+|description={{ module.description }}
+|owner(s)={{ people_entry(module.owners) -}}
+|peer(s)={{ people_entry(module.peers) -}}
+|includes={{ module.includes|join(", ") }}
+|excludes={{ module.excludes|join(", ") }}
+{%- if module.meta.group -%}
+|group={{ module.meta.group }}
+{% endif %}
+{%- if module.meta.url -%}
+|url={{ module.meta.group }}
+{% endif %}
+{%- if module.meta.components -%}
+|url={{ module.meta.components }}
+{% endif %}
+{% raw %}}}{% endraw %}
+{% endmacro %}
+
+
+{{ directory.description or "No directory description provided." }}
+
+{%- for module in directory.modules -%}
+{{ module_entry(module) }}
+{% if module.submodules %}
+== Submodules ==
+{% for submodule in module.submodules %}
+{{ module_entry(submodule) }}
+{% endfor %}
+{% endif %}
+{% endfor %}
+"""
+
+
+class Directory:
+    """Mots directory and path index."""
+
+    def __init__(self, config: FileConfig):
+        self.config_handle = config
+        self.config_handle.load()
+
+        self.repo_path = self.config_handle.repo_path
+        self.modules = [
+            Module(repo_path=self.repo_path, **m)
+            for m in self.config_handle.config["modules"]
+        ]
+
+        self.modules_by_machine_name = {}
+        for module in self.modules:
+            self.modules_by_machine_name[module.machine_name] = module
+
+        self.index = None
+        self.people = None
+
+    def load(self, full_paths: bool = False, query_bmo=True):
+        """Load all paths in each module and put them in index."""
+        self.index = defaultdict(list)
+
+        if full_paths:
+            # Make sure all existing paths have an entry in the index.
+            logger.debug("Loading paths into index...")
+            for path in self.repo_path.glob("**/*"):
+                if path.relative_to(self.repo_path).parts[0] in (".hg", ".git"):
+                    continue
+                self.index[path] = list()
```

Comment:
Is it OK that this will include potentially-large ignored directories, like `obj-*`?

Reply:
Filed [a bug](https://bugzilla.mozilla.org/show_bug.cgi?id=1753713) for this.